dask_sql/physical/rel/logical/optimize.py

```diff
@@ -0,0 +1,349 @@
+import logging
+import operator
+
+import dask.dataframe as dd
+import numpy as np
+from dask.blockwise import Blockwise
+from dask.highlevelgraph import HighLevelGraph
+from dask.layers import DataFrameIOLayer
+from dask.utils import M, apply, is_arraylike
+
+logger = logging.getLogger(__name__)
+
+
+def attempt_predicate_pushdown(ddf: dd.DataFrame) -> dd.DataFrame:
+    """Use graph information to update IO-level filters
+
+    The original `ddf` will be returned if/when the
+    predicate-pushdown optimization fails.
+
+    This is a special optimization that must be called
+    eagerly on a DataFrame collection when filters are
+    applied. The "eager" requirement for this optimization
+    is due to the fact that `npartitions` and `divisions`
+    may change when this optimization is applied (invalidating
+    npartition/divisions-specific logic in following Layers).
+    """
+
+    # Check that we have a supported `ddf` object
+    if not isinstance(ddf, dd.DataFrame):
+        raise ValueError(
+            f"Predicate pushdown optimization skipped. Type {type(ddf)} "
+            f"does not support predicate pushdown."
+        )
+    elif not isinstance(ddf.dask, HighLevelGraph):
+        logger.warning(
+            f"Predicate pushdown optimization skipped. Graph must be "
+            f"a HighLevelGraph object (got {type(ddf.dask)})."
+        )
+        return ddf
+
+    # We were able to extract a DNF filter expression.
+    # Check that we have a single IO layer with `filters` support
+    io_layer = []
+    for k, v in ddf.dask.layers.items():
+        if isinstance(v, DataFrameIOLayer):
+            io_layer.append(k)
+            if (
+                "filters" not in v.creation_info.get("kwargs", {})
+                or v.creation_info["kwargs"]["filters"] is not None
```

Comment:
In cases where this was created from  a `dd.read_csv` call, `v.creation_info` returns a noneType

Reply:
Ah, oops - I just introduced this bug today when I moved this IO-layer check to happen first.